dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/ClientCore.cs

```diff
@@ -132,15 +136,39 @@ internal ClientCore(ILogger? logger = null)
 
         var options = CreateCompletionsOptions(text, textExecutionSettings, this.DeploymentOrModelName);
 
-        var responseData = (await RunRequestAsync(() => this.Client.GetCompletionsAsync(options, cancellationToken)).ConfigureAwait(false)).Value;
-        if (responseData.Choices.Count == 0)
+        Completions? responseData = null;
+        List<TextContent> responseContent;
+        using (var activity = ModelDiagnostics.StartCompletionActivity(this.Endpoint, this.DeploymentOrModelName, ModelProvider, text, executionSettings))
         {
-            throw new KernelException("Text completions not found");
+            try
+            {
+                responseData = (await RunRequestAsync(() => this.Client.GetCompletionsAsync(options, cancellationToken)).ConfigureAwait(false)).Value;
+                if (responseData.Choices.Count == 0)
+                {
+                    throw new KernelException("Text completions not found");
+                }
+            }
+            catch (Exception ex)
+            {
+                activity?.SetError(ex);
+                if (responseData != null)
+                {
+                    // Capture available metadata even if the operation failed.
+                    activity?
+                        .SetResponseId(responseData.Id)
+                        .SetPromptTokenUsage(responseData.Usage.PromptTokens)
+                        .SetCompletionTokenUsage(responseData.Usage.CompletionTokens);
+                }
+                throw;
+            }
+
+            responseContent = responseData.Choices.Select(choice => new TextContent(choice.Text, this.DeploymentOrModelName, choice, Encoding.UTF8, GetTextChoiceMetadata(responseData, choice))).ToList();
+            activity?.SetCompletionResponse(responseContent, responseData.Usage.PromptTokens, responseData.Usage.CompletionTokens);
```

Comment:
Presumably we still want to set information on the activity about token usage and the like even if there's an error (like no choices, though I'm not sure in what situation that could happen)

Reply:
Right, it will attempt to set the token usage even when there are no choices. Please see the catch block above.